File: 0064-minimum-path-sum/0064-minimum-path-sum.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
let dp;
let matrix;
var minPathSum = function(grid) {
    matrix = grid;

    return f_bu()
};

function f_bu(){
    let n = matrix.length;
    let m = matrix[0].length;

    dp = new Array(205);
    for(let i = 0; i<205; i++){
        dp[i] = new Array(205);
    }

    dp[0][0] = matrix[0][0];

   // fill first row

   for(let col = 1; col<m; col++){
        dp[0][col] = matrix[0][col] + dp[0][col-1]
   }

   // fill first col

   for(let row = 1; row<n; row++){
        dp[row][0] = matrix[row][0] + dp[row -1][0];
   }

   for(let row = 1; row<n; row++){
        for(let col = 1; col<m; col++){
            dp[row][col] = matrix[row][col] + Math.min(dp[row-1][col], dp[row][col-1]);
        }
   }

    return dp[n-1][m-1];
}

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public static int[][] dp;
    public static int[][] matrix;

    public static int f_bu(){
        dp = new int[205][205];
        int m = matrix.length;
        int n = matrix[0].length;

        dp[0][0] = matrix[0][0];

        // first row;
        for(int j = 1; j< n; j++){
            dp[0][j] = matrix[0][j] + dp[0][j-1];
        }

        // first col 
        for(int i = 1; i< m; i++){
            dp[i][0] = matrix[i][0] + dp[i-1][0];
        }

        for(int i = 1; i < m; i++){
            for(int j = 1; j < n; j++){
                dp[i][j] = matrix[i][j] + Math.min(dp[i-1][j], dp[i][j-1]);
            }
        }

        return dp[m-1][n-1];
    }

    public int minPathSum(int[][] grid) {
        matrix = grid;

        return f_bu();
    }
}